[Sig Events] Add architecture and operator docs (#7497)
elastic/docs-content: solutions/observability/streams/significant-events/knowledge-indicators.md

  - id: elastic-stack
---

# Knowledge Indicators [sig-events-ki]

Knowledge Indicators (KIs) are structured facts that Elastic extracts from your raw log data automatically without requiring schemas, service catalogs, or manual configuration. When you run extraction against a log stream, Elastic analyzes the raw data and returns facts about your environment: which services are running, the underlying infrastructure they rely on, how they depend on each other, and the log schemas they use.

> Rather than a static configuration, this knowledge accumulates over time, automatically expires when a service disappears, and feeds directly into downstream capabilities like the Significant Events detection pipeline, topology maps, AI agent investigations, and dashboards.

Reviewer: This would allow you to the remove the next sentence, which is kinda floating in between the paragraphs.

```suggestion
Rather than a static configuration, this knowledge accumulates over time, automatically expires when a service disappears, and feeds directly into downstream capabilities like the [Significant Events detection pipeline](../significant-events/how-it-works.md), topology maps, AI agent investigations, and dashboards.
```

> KIs feed the Significant Events pipeline. See [How Significant Events works](../significant-events/how-it-works.md) for how KIs connect to detection, discovery, and triage.

Reviewer: ```suggestion
```

To access Knowledge Indicators, open **Significant Events** from the Streams main page and select the **Knowledge Indicators** tab.

:::{admonition} Requirements
To use this feature, you need:

> - A [Generative AI connector](kibana://reference/connectors-kibana/gen-ai-connectors.md), which can incur additional costs.

Reviewer: ```suggestion
- A [Generative AI connector](kibana://reference/connectors-kibana/gen-ai-connectors.md).
```

Continuous extraction
:   When enabled, continuous extraction runs automatically on managed streams at the interval you configure. Continuous extraction is off by default. To enable it:

    1. From the **Streams** main page, select **Significant Events** → **Settings**.
    1. Under **Continuous KI extraction**, turn on **Enable continuous KI extraction**.
    1. Set the **Extraction interval** in hours, and list any **Excluded streams** to skip during continuous extraction.

## How KI extraction works [sig-events-ki-extraction]

The extraction pipeline samples a small batch of logs from a stream and processes them through a combination of large language model (LLM) analysis and deterministic code generators. It accumulates its findings across multiple iterations, entirely configuration-free.

The pipeline runs up to five iterations. Each iteration fetches a batch of up to 20 documents assembled from three prioritized buckets:

| Bucket | Share | Strategy |
|---|---|---|

> | Entity-filtered | 40% | Random sample that excludes documents matching already-discovered feature filters, steering toward undiscovered patterns |

Reviewer: What's a feature filter? I don't think I've heard of that term before.

Reviewer: Oh, it's defined further down. It'd be good to define it here first, just so users aren't wondering what the term means.

```suggestion
| Entity-filtered | 40% | Random sample that excludes documents matching already-discovered feature filters (features excluded by users), steering toward undiscovered patterns |
```

- **Dependency**: `fraud-check` → `policy-lookup` (through an outbound HTTP call)
- **Technology**: Java, Spring Boot

### LLM analysis [sig-events-ki-llm]

Sampled documents are sent to an LLM that identifies the following feature types:

| Type | What it captures |
|------|-----------------|
| Entity | Distinct system components: services, applications, jobs |
| Infrastructure | Environment context: {{k8s}}, cloud provider, OS |
| Technology | Languages, frameworks, libraries, databases |
| Dependency | Relationships between components |
| Schema | Log format conventions: {{product.ecs}}, OTel, custom |

> Every feature must include stable identifying properties and cite direct evidence from the sampled logs. The LLM assigns a confidence score from 0–100 for each KI. The pipeline also tracks features excluded by users (false positives) and carries them forward to prevent re-identification in future runs.

Reviewer: In a future iteration, it'd be useful to have a section explaining the confidence score ranges. For now, it'd be sufficient to explain the 0-100 scale, for example:

```suggestion
Every feature must include stable identifying properties and cite direct evidence from the sampled logs. The LLM assigns a confidence score from 0–100 (where 0 is the lowest confidence and 100 is the highest) for each KI. The pipeline also tracks features excluded by users (false positives) and carries them forward to prevent re-identification in future runs.
```

### Deterministic generators [sig-events-ki-generators]

In parallel with LLM analysis, a set of deterministic code-based generators independently analyze the data to produce statistical summaries, log samples, pattern clusters, and error-specific features. Because these are computed rather than inferred, they always receive a confidence score of 100.

> ### Merging results [sig-events-ki-merge]

Reviewer: A possible alternative that aligns with the format the other section titles follow.

```suggestion
### Results handling [sig-events-ki-merge]
```

> LLM results and computed features are merged and deduplicated. Known KIs reuse their existing UUIDs, new discoveries get fresh ones, and the pipeline drops user-excluded features. Surviving KIs are saved with an active status and an expiration date set seven days out.

Reviewer 2: issue (blocking): expiration is 30 days, not seven — and this line contradicts your own lifecycle section.

`feature_ttl_days: 30` (`tuning_config.ts:56`), and the keep-alive rolls it forward by another 30. Your "KI lifecycle and maintenance" section correctly says "auto-expire after 30 days," and `how-it-works.md` says 30 too. This one sentence is the outlier.

```suggestion
LLM results and computed features are merged and deduplicated. Known KIs reuse their existing UUIDs, new discoveries get fresh ones, and the pipeline drops user-excluded features. Surviving KIs are saved with an active status and an expiration date set 30 days out.
```

}
```

Query KIs are generated in two {{esql}} forms depending on what the LLM determines is most useful for detection:

**Match queries** filter for specific log events — errors, exceptions, failure messages:

```esql
FROM logs-mystream,logs-mystream.* METADATA _id, _source
| WHERE message : "connection refused" AND service.name == "api_gateway"
```

**Stats queries** aggregate metrics over time — rates, counts, or derived values — useful when the signal is a change in volume rather than the presence of a specific event:

```esql
FROM logs-mystream,logs-mystream.*
| STATS error_count = COUNT(*) WHERE level == "ERROR" BY service.name, @timestamp
```

> ### Downstream path: from query KIs to alerting rules [sig-events-ki-downstream]

Reviewer: ```suggestion
### Downstream path: From query KIs to alerting rules [sig-events-ki-downstream]
```

> When you promote a query KI, it becomes a {{kib}} alerting rule. Each promoted rule runs its {{esql}} query on a per-rule schedule and writes results to `.rule-events`. The number of promoted query KIs is the main driver of alerting query load on your cluster.

Reviewer: Making some tweaks based on info from the program repo. 

```suggestion
When you promote a query KI, it becomes an alerting rule that runs its {{esql}} query on a schedule and writes a match count to `.rule-events` for each time bucket. The number of promoted query KIs, and how often each one executes, is the main driver of alerting query load on your cluster.
```

> When you promote a query KI, it becomes a {{kib}} alerting rule. Each promoted rule runs its {{esql}} query on a per-rule schedule and writes results to `.rule-events`. The number of promoted query KIs is the main driver of alerting query load on your cluster.

Reviewer 2: nitpick (non-blocking): true only for MATCH queries.

STATS queries can't be rule-backed yet (`canQueryBeRuleBacked` returns false for STATS; installing one throws, gated on #265778). So a STATS query KI won't become an alerting rule today. Might be worth a half-sentence so nobody's surprised when their STATS KI never fires.

> The Significant Events pipeline picks up from there: a detection workflow runs the `change_point` aggregation over alert firing patterns and writes significant transitions to `.significant_events-detections`. The discovery workflow then processes those detections with an AI agent.

Reviewer: ```suggestion
The Significant Events pipeline picks up from there. A Detection workflow runs change point aggregation over alert generation patterns and writes documents to `.significant_events-detections` when statistically significant shifts are found. A Discovery workflow then uses an LLM to process those documents.
```

## Continuous extraction [sig-events-ki-continuous]

> When continuous extraction is enabled, a {{kib}} Workflow runs every 35 minutes and processes up to five eligible streams per run. A stream is eligible if:

Reviewer: ```suggestion
When continuous extraction is enabled, a workflow runs every 35 minutes and processes up to five eligible streams per run. A stream is eligible if:
```

- It is a wired or classic stream

> - It does not match any configured exclusion glob pattern

Reviewer 2: issue (blocking): "does not match any configured exclusion glob pattern" describes a setting that was removed.

There is no exclusion-glob mechanism. Stream scoping is done by *inclusion* via `observability:streamsSigEventsIndexPatterns`, and the old `excludedStreamPatterns` workflow input was deleted — there's even a test asserting it's gone (`continuous_onboarding_workflow.test.ts`: "does not declare the removed excludedStreamPatterns input"). Two other nuances: query streams are also eligible (not just wired/classic), and non-query streams must *match* the configured index patterns. Suggest:

```suggestion
- It is a wired, classic, or query stream that matches the configured Significant Events index patterns
- No feature identification is currently running for it
```

- No feature identification task is currently running for it
- Enough time has elapsed since its last extraction (controlled by the **Extraction interval** setting, default 12 hours)

Streams that have never been processed are always prioritized. Among remaining candidates, streams with the oldest last-completed extraction run are processed first.

The continuous extraction workflow has a 34-minute timeout (one minute shorter than the 35-minute schedule, to prevent overlapping runs). If multiple runs would overlap, excess triggers are silently dropped.

> Toggling continuous extraction off cancels any in-flight extraction tasks and force-deletes the workflow. Re-enabling it creates a fresh workflow from the latest definition. The workflow state is not preserved between enable/disable cycles.

Reviewer: ```suggestion
Toggling continuous extraction off cancels any in-flight extraction tasks and force-deletes the workflow. Re-enabling it creates a fresh workflow from the latest definition. The workflow state is not preserved between enable or disable cycles.
```

> Toggling continuous extraction off cancels any in-flight extraction tasks and force-deletes the workflow. Re-enabling it creates a fresh workflow from the latest definition. The workflow state is not preserved between enable/disable cycles.

Reviewer 2: suggestion: the disable/re-enable description doesn't match the current managed workflow.

For the managed continuous workflow, disabling *disables* it (sets `enabled: false`) and cancels in-flight onboarding; it does not force-delete it. Only the *legacy* pre-migration workflow gets force-deleted. Re-enabling flips `enabled: true` on the same workflow — it's `enablement: 'restorable'`, so it's not recreated from scratch and there's no "fresh workflow." Suggest:

```suggestion
Toggling continuous extraction off cancels any in-flight extraction tasks and disables the workflow. Re-enabling turns it back on; already-extracted KIs are preserved throughout.
```

### Continuous extraction settings

| Setting | Default | Description |
|---|---|---|
| `observability:streamsContinuousKiExtractionEnabled` | `false` | Enables or disables continuous extraction |
| `observability:streamsContinuousKiExtractionIntervalHours` | `12` | Minimum hours between extraction runs for a single stream |

> | `observability:streamsContinuousKiExtractionExcludedStreamPatterns` | `""` | Comma-separated glob patterns for streams to skip |

Reviewer 2: issue (blocking): `observability:streamsContinuousKiExtractionExcludedStreamPatterns` doesn't exist.

I grepped the whole repo — no setting ID, no registration, nothing. The other two rows (`...Enabled` default false, `...IntervalHours` default 12) are correct. Drop the excluded-patterns row. If you want to document stream scoping, point at `observability:streamsSigEventsIndexPatterns` (the inclusion patterns on the Settings page) instead.

> These settings are only modifiable through the Significant Events Settings page, not through {{kib}} Advanced Settings.

Reviewer: Statement stands without the last part.

```suggestion
These settings are only modifiable through the Significant Events Settings page.
```
